src/shared/test/integration/inter-frame-communication-test.js

```diff
@@ -0,0 +1,178 @@
+    constructor(...args) {
+      super(...args);
+      destroyables.push(this);
+    }
+  }
+
+  class PortProvider extends PortProvider_ {
+    constructor(...args) {
+      super(...args);
+      destroyables.push(this);
+    }
+  }
+
+  beforeEach(() => {
+    destroyables = [];
+  });
+
+  afterEach(() => {
+    destroyables.forEach(instance => instance.destroy());
+  });
+
+  it('enables the communication between guest-host', async () => {
+    let done;
+    const promise = new Promise(resolve => (done = resolve));
+
+    // guest frame
+    const portFinder = new PortFinder({
+      hostFrame: window,
+      source: 'guest',
+    });
+    const hostRPC = new PortRPC();
+
+    portFinder.discover('host').then(port => {
+      hostRPC.connect(port);
+      hostRPC.call('ping', response => {
+        assert.equal(response, 'pong');
+        done();
+      });
+    });
+
+    await delay(10); // add some realism
+
+    // host frame
+    const portProvider = new PortProvider(window.location.origin);
+    portProvider.listen();
+    const guestRPC = new PortRPC();
+
+    // Register RPC method *before* connection
+    guestRPC.on('ping', cb => cb('pong'));
+
+    portProvider.on('frameConnected', (source, port) => {
+      if (source === 'guest') {
+        guestRPC.connect(port);
+      }
+    });
+
+    return promise;
+  });
+
+  it('enables the communication between guest-sidebar', async () => {
+    let done;
+    const promise = new Promise(resolve => (done = resolve));
+
+    // guest frame;
+    const portFinder1 = new PortFinder({
+      hostFrame: window,
+      source: 'guest',
+    });
+    const sidebarRPC = new PortRPC();
+
+    portFinder1.discover('sidebar').then(port => {
+      sidebarRPC.connect(port);
+      sidebarRPC.call('ping', response => {
+        assert.equal(response, 'pong');
+        done();
+      });
+    });
+
+    await delay(10); // add some realism
+
+    // sidebar frame
+    const portFinder2 = new PortFinder({
+      hostFrame: window,
+      source: 'sidebar',
+    });
+    const guestRPC = new PortRPC();
+
+    // Register RPC method *before* connection
+    guestRPC.on('ping', cb => cb('pong'));
+
+    const listenerCollection = new ListenerCollection();
+    portFinder2.discover('host').then(port => {
+      listenerCollection.add(port, 'message', ({ data, ports }) => {
+        if (
+          isMessageEqual(data, {
+            frame1: 'guest',
+            frame2: 'sidebar',
+            type: 'offer',
+          })
+        ) {
+          guestRPC.connect(ports[0]);
+        }
+      });
+      port.start(); // `start` method would be triggered by `hostRPC.connect(port)`
+    });
+
+    // host frame
+    const portProvider = new PortProvider(window.location.origin);
+    portProvider.listen();
+
+    return promise;
+  });
+
+  it('enables the communication between sidebar-host', async () => {
+    let done;
+    const promise = new Promise(resolve => (done = resolve));
```

Comment:
If you are going to use this pattern, I'd suggest to let `done` be the promise as it is a name of a state (similar to `loaded` or `completed`), and call the resolver `resolveDone`. Alternatively, skip the resolver variable and just use the promise callback:

```js
const done = new Promise(resolve => {
   ...
   hostRPC.call('ping', response => { ...; resolve(); });
});
```